jetty-server/src/test/java/org/eclipse/jetty/server/BlockingContentProducerTest.java

```diff
@@ -0,0 +1,320 @@
+        buffers[2] = gzipByteBuffer(uncompressedBuffers[2]);
+
+        AtomicReference<ContentProducer> ref = new AtomicReference<>();
+        ArrayDelayedHttpChannel httpChannel = new ArrayDelayedHttpChannel(buffers, new HttpInput.EofContent(), scheduledExecutorService, () -> ref.get().wakeup());
+        ContentProducer contentProducer = new BlockingContentProducer(httpChannel, new AsyncContentProducer(httpChannel));
+        ref.set(contentProducer);
+        contentProducer.setInterceptor(new GzipHttpInputInterceptor(inflaterPool, new ArrayByteBufferPool(1, 1, 2), 1));
+
+        Throwable error = readAndAssertContent(totalContentBytesCount, originalContentString, totalContentBytesCount + 1, contentProducer);
+        assertThat(error, nullValue());
+    }
+
+    @Test
+    public void testBlockingContentProducerGzipInterceptorWithError()
+    {
+        ByteBuffer[] uncompressedBuffers = new ByteBuffer[3];
+        uncompressedBuffers[0] = ByteBuffer.wrap("1 hello 1".getBytes(StandardCharsets.ISO_8859_1));
+        uncompressedBuffers[1] = ByteBuffer.wrap("2 howdy 2".getBytes(StandardCharsets.ISO_8859_1));
+        uncompressedBuffers[2] = ByteBuffer.wrap("3 hey ya 3".getBytes(StandardCharsets.ISO_8859_1));
+        final int totalContentBytesCount = countRemaining(uncompressedBuffers);
+        final String originalContentString = asString(uncompressedBuffers);
+        final Throwable expectedError = new Throwable("HttpInput idle timeout");
+
+        ByteBuffer[] buffers = new ByteBuffer[3];
+        buffers[0] = gzipByteBuffer(uncompressedBuffers[0]);
+        buffers[1] = gzipByteBuffer(uncompressedBuffers[1]);
+        buffers[2] = gzipByteBuffer(uncompressedBuffers[2]);
+
+        AtomicReference<ContentProducer> ref = new AtomicReference<>();
+        ArrayDelayedHttpChannel httpChannel = new ArrayDelayedHttpChannel(buffers, new HttpInput.ErrorContent(expectedError), scheduledExecutorService, () -> ref.get().wakeup());
+        ContentProducer contentProducer = new BlockingContentProducer(httpChannel, new AsyncContentProducer(httpChannel));
+        ref.set(contentProducer);
+        contentProducer.setInterceptor(new GzipHttpInputInterceptor(inflaterPool, new ArrayByteBufferPool(1, 1, 2), 32));
+
+        Throwable error = readAndAssertContent(totalContentBytesCount, originalContentString, buffers.length + 1, contentProducer);
+        assertThat(error, is(expectedError));
+    }
+
+    private Throwable readAndAssertContent(int totalContentBytesCount, String originalContentString, int totalContentCount, ContentProducer contentProducer)
+    {
+        int readBytes = 0;
+        int nextContentCount = 0;
+        String consumedString = "";
+        Throwable error = null;
+        while (true)
+        {
+            HttpInput.Content content = contentProducer.nextContent();
+            nextContentCount++;
+
+            if (content.isSpecial())
+            {
+                if (content.isEof())
```

Comment:
"eof" implies "special"?

Reply:
Yes, any content that does not wrap a `ByteBuffer` is special.